cli/src/tasks/migrate.ts

```diff
@@ -0,0 +1,590 @@
+          androidxEspressoCoreVersion: '3.4.0',
+          cordovaAndroidVersion: '10.1.1',
+        };
+        for (const variable of Object.keys(variables)) {
+          if (
+            !updateFile(
+              join('android', 'variables.gradle'),
+              `${variable} = '`,
+              `'`,
+              variables[variable].toString(),
+              true,
+            )
+          ) {
+            updateFile(
+              join('android', 'variables.gradle'),
+              `${variable} = `,
+              `\n`,
+              variables[variable].toString(),
+              true,
+            );
+          }
+        }
+      }
+
+      // Ran Cap Sync
+      await getCommandOutput('npx', ['cap', 'sync']);
+
+      writeBreakingChanges();
+      logSuccess(
+        `Migration to Capacitor ${coreVersion} is complete. Run and test your app!`,
+      );
+    } catch (err) {
+      fatal(`Failed to migrate: ${err}`);
+    }
+  } else {
+    fatal(`User canceled migration.`);
+  }
+}
+
+function daysUntil(date_1: Date) {
+  const date_2 = new Date();
+  const difference = date_1.getTime() - date_2.getTime();
+  return Math.ceil(difference / (1000 * 3600 * 24));
+}
+
+async function installLatestNPMLibs() {
+  const result: string[] = [];
+  for (const lib of libs) {
+    if (allDependencies[lib]) {
+      result.push(`${lib}@${coreVersion}`);
+    }
+  }
+
+  for (const plugin of plugins) {
+    if (allDependencies[plugin]) {
+      result.push(`${plugin}@${pluginVersion}`);
+    }
+  }
+
+  await getCommandOutput('npm', ['i', ...result]);
+}
+
+async function migrateStoragePluginToPreferences() {
+  if (allDependencies['@capacitor/storage']) {
+    await getCommandOutput('npm', ['uninstall', '@capacitor/storage']);
+    await getCommandOutput('npm', [
+      'i',
+      `@capacitor/preferences@${pluginVersion}`,
+    ]);
+    logger.info('Migrated @capacitor/storage to @capacitor/preferences.');
+  }
+}
+
+function writeBreakingChanges() {
+  const breaking = [
+    '@capacitor/storage',
+    '@capacitor/camera',
+    '@capacitor/push-notifications',
+    '@capacitor/local-notifications',
+  ];
+  const broken = [];
+  for (const lib of breaking) {
+    if (allDependencies[lib]) {
+      broken.push(lib);
+    }
+  }
+  if (broken.length > 0) {
+    logger.info(
+      `IMPORTANT: Review https://next.capacitorjs.com/docs/updating/4-0#plugins for breaking changes in these plugins that you use: ${broken.join(
+        ', ',
+      )}.`,
+    );
+  }
+  if (allDependencies['@capacitor/android']) {
+    logger.info(
+      'Warning: The Android Gradle plugin was updated and it requires Java 11 to run. You may need to select this in Android Studio.',
+    );
+  }
+}
+
+function updateAndroidManifest(filename: string) {
+  const txt = readFile(filename);
+  if (!txt) {
+    return;
+  }
+
+  // AndroidManifest.xml add attribute: <activity android:exported="true"
+  if (txt.includes('<activity android:exported="')) {
```

Comment:
the exported can be on any line of the activity tag, I personally put it on the last line and this doesn't detect it and duplicates it.